src/lib/broadcast.jl

```diff
@@ -232,23 +235,44 @@ end
 import ForwardDiff
 using ForwardDiff: Dual
 
-dual(x, p) = x
-dual(x::Real, p) = Dual(x, p)
-dual(x::Bool, p) = x
+
+# We do this because it ensures type stability so it compiles nicely on the gpu
+dual(x, i, N) = x
+dual(x::Bool, i, ::Val{N}) where {N} = x
+dual(x::Real, i, ::Val{N}) where {N} = Dual(x, ntuple(j-> i==j, Val(N)))
+# For complex since ForwardDiff.jl doesn't play nicely with complex numbers we
+# construct a Complex dual number and tag the real and imaginary parts separately
+function dual(x::Complex, i, ::Val{N}) where {N}
+    re_dual = Dual(real(x), ntuple(j->i==j, Val(2N)))
+    im_dual = Dual(imag(x), ntuple(j->(N+i)==j, Val(2N)))
+    return Complex(re_dual, im_dual)
+end
 
 function dual_function(f::F) where F
-  function (args::Vararg{Any,N}) where N
-    ds = map(args, ntuple(identity,Val(N))) do x, i
-      dual(x, ntuple(j -> i==j, Val(N)))
+    function (args::Vararg{Any,N}) where N
+      ds = map(args, ntuple(identity,Val(N))) do x, i
+        tmp = dual(x, i, Val(N))
```

Comment:
```suggestion
      ds = map(args, ntuple(identity, N)) do x, i
        tmp = dual(x, i, N)
```
Same as above. I believe all subsequent uses of `Val(N)` might be redundant as well.

Reply:
So it looks like I can get rid of almost every Val. However, if I get rid of this one the following test fails
https://github.com/FluxML/Zygote.jl/blob/6742644da91ad22e3ab7081c4d8b77eeeafe4e79/test/cuda.jl#L20-L28
due to a type instability. 

If I instead do this
```suggestion julia
      ds = map(args, ntuple(identity, N) do x, i
        tmp = dual(x, i, Val(N))
```
then everything works on 1.6 and 1.8 on my machine, so I think this Val needs to remain.